populus/compilation/compile.py

```diff
@@ -0,0 +1,89 @@
+import logging
+import os
+
+from .helpers import (
+    get_dir_source_paths,
+    post_process_compiled_contracts,
+    validate_compiled_contracts,
+)
+
+from .backends.solc_auto import (
+    get_compiler_backend_class_for_version
+)
+
+from populus.utils.functional import (
+    get_duplicates,
+)
+
+
+from populus.config.loading import (
+    load_global_config,
+)
+
+
+def _get_contract_key(contract_data):
+    return (
+        contract_data['source_path'],
+        contract_data['name'],
+    )
+
+
+def compile_dirs(dir_paths,
+                 global_config_path=None,
+                 import_remappings=None,
+                 compiler_version="auto"
+                 ):
+
+    logger = logging.getLogger('populus.compilation.helpers.compile_dir')
+    if import_remappings is None:
+        import_remappings = []
+
+    all_source_paths = []
+    for dir_path in set(dir_paths): # sliently ignore duplicates
+        contract_source_paths = get_dir_source_paths(dir_path)
+        logger.debug(
+            "Found %s source files: %s",
+            len(contract_source_paths),
+            ", ".join(contract_source_paths),
+        )
+        all_source_paths.extend(contract_source_paths)
```

Comment:
I dislike the mutability.  Can you extract this into a stand-alone function that uses the `eth_utils.to_tuple` decorator to generate this list.

```python
from eth_utils import to_tuple

def _collect_source_paths(dir_paths):
    for dir_path in dir_paths:
        for source_path in get_dir_source_paths(dir_path):
            yield source_path
```

Removes the mutability and cleans up the body of the function some.